### DISCUSSION TITLE
Configuration phase 2 (Client Profile configuration implementation)

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#1341] [Ideas]
What is the vision of the configuration?
---
- We would like to achieve a configurable register flow that starts from the **Navigation drawer** to the **Client profile**.
- This will enable us to configure the following 
    - The menu item for a register and when it's displayed on the navigation drawer. 
    - The register that opens after the menu above is clicked, the configuration of the data to be displayed on the register. 
    - The profile that is opened after the register row is clicked, should also define what is displayed on the profile and the menu actions to be performed on the profile. 
- Previous steps --> [Configuration phase 2 (Registers configuration implementation)](https://github.com/opensrp/fhircore/discussions/1340)

Client Profile Configuration
---
- To create a configuration that allows us to define a profile for specific verticals. These configs should also help configure the menus that are shown in every profile. These are the things we would what to nail down
    - Configurations to show or hide extra information 
    - Configurations to show or hide some cards e.g the Related Patients, ANC, Patient Medical History and services.  
       - Configurations to define the data to be displayed on every card to be displayed on the profile. 
    - Configure the following items on the profile menus. 
       - Configure what menus are the default for each profile e.g the Edit and Remove menus in each profile. 
       - Configure the extra menus shown on the profile. Identify what form actions they achieve and what conditions necessitate the display of the menus. 

**How to achieve the configuration**
- Family profile --> https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=5731%3A54488
- Patient Profile --> https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=5778%3A62393, https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=9860%3A75204

1. UI Items Identification
    - Header on the page topbar -- This is mostly shown when scrolling. The Title/Name moves up to the top bar
    - Demographics card
        -  Title/Name row
        -  Important attributes row
            - Roles row e.g showing whether a patient is a household head
            - ID display
        -  Assigned CarePlans row
        - Other information cards
            - Subheading row
            - Data/Text row 
        -  Special attribute display i.e https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=904%3A8274
            - Card row
               -  Column -- Title & Value
    - Service cards
        - Top row
           - Title (Left screen aligned)
           - Button (right screen aligned) 
        - Body
            - Register of different kinds -- This could utilise the Register configuration work.  
    - Floating  button
        - Icon
        - Label
        - Extra button functionality.  

2. Data load specification
    - Header on the topbar -- THis will be done later. 
    - Demographics card
       -  All the text will use the `COMPOUND_TEXT` component we already have in the configuration of the registers. 
       - The card holding the gender, age & dob will have a card that has `visibility` toggles. 
           - We will use the `COMPOUND_TEXT` component to display the text on the text views
    - Service cards
       - The cards will have two sections. 
           - The header will have title text on the left and the see all text on the write. 
               - All the text pieces will use the `COMPOUND_TEXT` component. We will however update the component to have a click listener and be able to redirect to other screens. 
           - The body - this is where the rest of the information is shown. 
               - Most of these displays are lists of different things e.g tasks, conditions, test results e.t.c. 
               - For this, we reuse the register configuration used for the register pieces
    - Floating button
    - Profile menus. 

 **Configuration sample** 
```json
{
  "appId": "app",
  "configType": "profile",
  "id": "householdProfile",
  "fhirResource": {
    "baseResource": {
      "resource": "Group",
      "dataQueries": [
        {
          "id": "householdQueryByType",
          "filterType": "TOKEN",
          "key": "type",
          "valueType": "CODING",
          "valueCoding": {
            "system": "http://hl7.org/fhir/group-type",
            "code": "person"
          }
        },
        {
          "id": "householdQueryByCode",
          "filterType": "TOKEN",
          "key": "code",
          "valueType": "CODEABLECONCEPT",
          "valueCoding": {
            "system": "https://www.snomed.org",
            "code": "35359004"
          }
        }
      ]
    },
    "relatedResources": [
      {
        "resource": "Patient",
        "fhirPathExpression": "Group.member.entity",
        "relatedResources": [
          {
            "resource": "Condition",
            "searchParameter": "subject"
          },
          {
            "resource": "CarePlan",
            "searchParameter": "subject"
          },
          {
            "resource": "Task",
            "searchParameter": "subject"
          }
        ]
      },
      {
        "resource": "CarePlan",
        "searchParameter": "subject",
        "dataQueries": [
          {
            "id": "filterHouseholdCarePlans",
            "filterType": "TOKEN",
            "key": "_tag",
            "valueType": "CODING",
            "valueCoding": {
              "system": "https://www.snomed.org",
              "code": "35359004"
            }
          }
        ]
      }
    ]
  },
  "rules": [
    {
      "name": "familyId",
      "condition": "true",
      "actions": [
        "data.put('familyId', fhirPath.extractValue(Task, 'Task.description'))"
      ]
    }
  ],
  "overFlowMenuItems": [
    {
      "title" : "Family details",
      "titleColor": "#000000",
      "visible": "true",
      "actions": [
        {
          "trigger": "ON_CLICK",
          "workflow": "LAUNCH_PROFILE"
        }
      ]
    },
    {
      "title" : "Change family head",
      "titleColor": "#000000",
      "visible": "true",
      "actions": [
        {
          "trigger": "ON_CLICK",
          "workflow": "LAUNCH_PROFILE"
        }
      ]
    },
    {
      "title" : "Family activity",
      "titleColor": "#000000",
      "visible": "true",
      "actions": [
        {
          "trigger": "ON_CLICK",
          "workflow": "LAUNCH_PROFILE"
        }
      ]
    },
    {
      "title" : "View past encounters",
      "titleColor": "#000000",
      "visible": "true",
      "actions": [
        {
          "trigger": "ON_CLICK",
          "workflow": "LAUNCH_REGISTER",
          "id": ""
        }
      ]
    },
    {
      "title" : "Remove family",
      "titleColor": "#000000",
      "visible": "true",
      "actions": [
        {
          "trigger": "ON_CLICK"
        }
      ]
    }
  ]
}
```
Next steps --> [Configuration phase 2 (Practitioner Profile configuration implementation)](https://github.com/opensrp/fhircore/discussions/1342)

**Comment by pld** [comment]
@HenryRae do we have a profile design system for reference?

**Reply by dubdabasoduba** [reply]
@pld 
- Family profile --> https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=5731%3A54488
- Patient Profile --> https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=5778%3A62393, https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=9860%3A75204

**Reply by pld** [reply]
OK cool, thanks, listing out the different components we'll need for these example profiles:

1. household member list (feels like a register)
2. add thing button (like the add family member)
3. demographic info
4. visit button list (feels like a weird register)
5. test results list (feels like a register)
6. guardian details list (feels like a register)

do we have an example of the profiles in goldsmith, and what supervisors see in practitioner profiles? we can add that to the component list if there's anything not covered in the above

**Reply by HenryRae** [reply]
sorry just seeing this @pld. Let me know if you still need help in this.

**Comment by pld** [comment]
Chatted briefly Vitor and Jing about this, FHIRPath could give us the the queries into a single resource, but to query related resources we'd need CQL or maybe something called XFHIRQuery, which doesn't seem to have docs.

There's also potential work to extend the Android FHIR SDK's search interface that we could use, maybe we could use that interface now? Not sure
